src/transformStateWithClones.js

```diff
@@ -7,7 +7,34 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const statesArray = [];
+
+  let currentState = { ...state };
+
+  actions.forEach((action) => {
+    switch (action.type) {
+      case 'clear':
+        currentState = {};
+        break;
+
+      case 'addProperties':
+        currentState = { ...currentState, ...action.extraData };
+        break;
+
+      case 'removeProperties':
+        action.keysToRemove.forEach((key) => {
+          delete currentState[key];
+        });
```

Comment:
Using `delete` on the `currentState` object directly mutates the object, which goes against the task's requirement to not mutate the state. Instead, consider creating a new object that excludes the keys to remove. You can achieve this by using `Object.keys`, `filter`, and `reduce` to construct a new object without the specified keys.